connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignor.java

```diff
@@ -753,6 +678,146 @@ private ConnectorsAndTasks assignment(Map<String, ConnectorsAndTasks> memberAssi
+                "connector",
+                configured.connectors().size(),
+                workers,
+                WorkerLoad::connectors
+        );
+        Map<String, Set<ConnectorTaskId>> taskRevocations = loadBalancingRevocations(
+                "task",
+                configured.tasks().size(),
+                workers,
+                WorkerLoad::tasks
+        );
+
+        connectorRevocations.forEach((worker, revoked) ->
+                result.computeIfAbsent(worker, w -> new ConnectorsAndTasks.Builder()).addConnectors(revoked)
+        );
+        taskRevocations.forEach((worker, revoked) ->
+                result.computeIfAbsent(worker, w -> new ConnectorsAndTasks.Builder()).addTasks(revoked)
+        );
+
+        return buildAll(result);
+    }
+
+    private <E> Map<String, Set<E>> loadBalancingRevocations(
+            String allocatedResourceName,
+            int totalToAllocate,
+            Collection<WorkerLoad> workers,
+            Function<WorkerLoad, Collection<E>> workerAllocation
+    ) {
+        int totalWorkers = workers.size();
+        // The minimum instances of this resource that should be assigned to each worker
+        int minAllocatedPerWorker = totalToAllocate / totalWorkers;
+        // How many workers are going to have to be allocated exactly one extra instance
+        // (since the total number to allocate may not be a perfect multiple of the number of workers)
+        int extrasToAllocate = totalToAllocate % totalWorkers;
+        // Useful function to determine exactly how many instances of the resource a given worker is currently allocated
+        Function<WorkerLoad, Integer> workerAllocationSize = workerAllocation.andThen(Collection::size);
+
+        long workersAllocatedMinimum = workers.stream()
+                .map(workerAllocationSize)
+                .filter(n -> n == minAllocatedPerWorker)
+                .count();
+        long workersAllocatedSingleExtra = workers.stream()
+                .map(workerAllocationSize)
+                .filter(n -> n == minAllocatedPerWorker + 1)
+                .count();
+        if (workersAllocatedSingleExtra == extrasToAllocate
+                && workersAllocatedMinimum + workersAllocatedSingleExtra == totalWorkers) {
+            log.trace(
+                    "No load-balancing {} revocations required; the current allocations, when combined with any newly-created {}, should be balanced",
+                    allocatedResourceName,
+                    allocatedResourceName
+            );
+            return Collections.emptyMap();
+        }
+
+        Map<String, Set<E>> result = new HashMap<>();
+        // How many workers we've allocated a single extra resource instance to
+        int allocatedExtras = 0;
+        for (WorkerLoad worker : workers) {
+            int currentAllocationSizeForWorker = workerAllocationSize.apply(worker);
+            if (currentAllocationSizeForWorker <= minAllocatedPerWorker) {
```

Comment:
I maybe missing something here, but the goal looks like to have each worker have either <m or m + 1> resource. If it has less that min, doesn't it become a candidate to receive some resources from others which have higher. Let's say it has 3 and minToAllocate is 5 and there's some other worker having 7, then 2 can be reduced from the latter and assigned to this one. I see below the logic to remove extra resources from workers.

Reply:
> If it has less that min, doesn't it become a candidate to receive some resources from others which have higher.

Exactly! The idea here (in the `performLoadBalancingRevocations` method) is just to do the revoking part of this logic. The other part--where we assign connectors and tasks to workers that may have fewer than the minimum--is handled later on, when we invoke `assignConnectors` and `assignTasks`. It may also take a follow-up rebalance for this to take place, since we have to wait for workers to stop running their revoked connectors and tasks before those can be assigned to different workers.

Reply:
Ok. I think this is where the other comment you made about being able to assign more connectors/tasks in a round makes more sense. This could mean we could converge faster.